ludwig/utils/defaults.py

```diff
@@ -260,29 +187,25 @@ def merge_with_defaults(config):
             raise ValueError("Stratify feature must be binary or category")
 
     # ===== Training =====
-    set_default_value(config, TRAINER, default_training_params)
-
-    for param, value in default_training_params.items():
-        set_default_value(config[TRAINER], param, value)
+    full_trainer_config, _ = load_config_with_kwargs(TrainerConfig, config[TRAINER] if TRAINER in config else {})
+    config[TRAINER] = asdict(full_trainer_config)
 
     set_default_value(
         config[TRAINER],
         "validation_metric",
         output_type_registry[config["output_features"][0][TYPE]].default_validation_metric,
     )
 
-    # ===== Training Optimizer =====
-    optimizer = config[TRAINER]["optimizer"]
-    default_optimizer_params = get_default_optimizer_params(optimizer[TYPE])
-    for param in default_optimizer_params:
-        set_default_value(optimizer, param, default_optimizer_params[param])
-
     # ===== Input Features =====
     for input_feature in config["input_features"]:
         get_from_registry(input_feature[TYPE], input_type_registry).populate_defaults(input_feature)
 
     # ===== Combiner =====
-    set_default_value(config, "combiner", {TYPE: default_combiner_type})
+    set_default_value(config, COMBINER, {TYPE: default_combiner_type})
+    full_combiner_config, _ = load_config_with_kwargs(
+        combiner_registry[config[COMBINER][TYPE]].get_schema_cls(), config[COMBINER]
```

Comment:
If this is working correctly locally, this is a moot point but because of the way the schemas and `combiners.py` are currently structured it is possible to import `combiner_registry` before any combiners have been registered. [See this thread](https://predibase.slack.com/archives/C03F25Q9LTB/p1654635842932799)

Might be safer to do an indirect import from `combiners.py` for now? (Though this may also cause a circular import...)

Reply:
This seems to be working locally, both in unit tests as well as a real training run.

Reply:
@justinxzhao Hello, could you please have a look at https://github.com/ludwig-ai/ludwig/issues/2448.